sqlite_utils/db.py

```diff
@@ -2966,102 +2999,128 @@ def build_insert_queries_and_params(
         replace,
         ignore,
     ):
-        # values is the list of insert data that is passed to the
-        # .execute() method - but some of them may be replaced by
-        # new primary keys if we are extracting any columns.
-        values = []
+        """
+        Given a list ``chunk`` of records that should be written to *this* table,
+        return a list of ``(sql, parameters)`` 2-tuples which, when executed in
+        order, perform the desired INSERT / UPSERT / REPLACE operation.
+        """
+        if (upsert or replace) and not pk:
+            raise PrimaryKeyRequired("UPSERT / REPLACE needs a primary key")
+
```

Comment:
Turns out this isn't needed at all.

Reply:
... actually it was, just not there. https://github.com/simonw/sqlite-utils/pull/653/commits/ac9f8208c7d24b92c983ea7d0f23545904a20e1d